core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableMergeJoin.java

```diff
@@ -101,28 +103,55 @@ public static boolean isMergeJoinSupported(JoinRelType joinType) {
     // Required collation keys can be subset or superset of merge join keys.
     RelCollation collation = required.getTrait(RelCollationTraitDef.INSTANCE);
     List<Integer> reqKeys = RelCollations.ordinals(collation);
-    ImmutableBitSet reqKeySet = ImmutableBitSet.of(reqKeys);
+    // need to copy reqKeys because it is not sortable.
+    reqKeys = new ArrayList<>(reqKeys);
+    List<Integer> leftKeys = immutableIntListToList(joinInfo.leftKeys, 0);
+    List<Integer> rightKeys = immutableIntListToList(joinInfo.rightKeys,
+        left.getRowType().getFieldCount());
+    List<Integer> rightKeysNotShifted = immutableIntListToList(joinInfo.rightKeys, 0);
 
+    ImmutableBitSet reqKeySet = ImmutableBitSet.of(reqKeys);
     ImmutableBitSet leftKeySet = ImmutableBitSet.of(joinInfo.leftKeys);
     ImmutableBitSet rightKeySet = ImmutableBitSet.of(joinInfo.rightKeys)
         .shift(left.getRowType().getFieldCount());
 
     Map<Integer, Integer> keyMap = new HashMap<>();
-    final int keyCount = leftKeySet.cardinality();
+    final int keyCount = leftKeys.size();
     for (int i = 0; i < keyCount; i++) {
       keyMap.put(joinInfo.leftKeys.get(i), joinInfo.rightKeys.get(i));
     }
     Mappings.TargetMapping mapping = Mappings.target(keyMap,
         left.getRowType().getFieldCount(),
         right.getRowType().getFieldCount());
 
-    // Only consider exact key match for now
+
     if (reqKeySet.equals(leftKeySet)) {
+      // if sort keys equal to left join keys, we can pass through all collations directly.
       RelCollation rightCollation = RexUtil.apply(mapping, collation);
       return Pair.of(
           required, ImmutableList.of(required,
           required.replace(rightCollation)));
+    } else if (isSubset(reqKeys, leftKeys)) {
+      // if sort keys are subset of left join keys, we can extend collations to make sure all join
+      // keys are sorted.
+      collation = extendCollation(collation, leftKeys);
+      RelCollation rightCollation = RexUtil.apply(mapping, collation);
+      return Pair.of(
+          required, ImmutableList.of(required.replace(collation),
+              required.replace(rightCollation)));
+    } else if (isPrefixOrderingNotRequired(leftKeys, reqKeys)
```

Comment:
You can use RelCollations.containsOrderless

Reply:
`isPrefixOrderingNotRequired` probably is the one I cannot use `RelCollations.containsOrderless`. It is designed for superset case. In which join keys should be 1) an orderless subset 2) prefix of required collations.

`RelCollations.containsOrderless` does not consider prefix requirement.

Alternatively, I am ok to move `isPrefixOrderingNotRequired` to `RelCollations` as a util function.

Reply:
For supset case, for example if required collation is [d, b, a ,c]
then join keys [d, b, a] is valid 
join keys [d,a,c] is not (if push down [d,b,a,c], it does not guarantee the ordering of [d,a,c]

Reply:
can you explain `RelCollations.containsOrderless does not consider prefix requirement.`?

Reply:
`containsOrderless` only consider if one is another's subset, for example:
[1, 2, 3] is [2, 3, 4, 1]'s subset
[3, 2, 1] is [2, 1, 3, 4] 's subset

The prefix requirement is, still use examples above, but apply sublist:
[2,3,4,1] -> [2,3,4], then containsOrderless([1,2,3]) = false.
[2,1,3,4] ->[2,1,3], then containsOrderless([3,2,1]) = true.

Reply:
In fact, I mis-read how `containsOrderless` is implemented. It actually meets the need here.

Have changed to `containsOrderless` and dropped unused code.